design/20241124-rename-bunde-to-clusterbundle.md

```diff
@@ -0,0 +1,252 @@
+
+### Goals
+
+- `Bundle` resource is renamed to `ClusterBundle`
+- `ClusterBundle` is created in the new `trust-manager.io` API group
+- Provide a smooth migration for users of trust-manager
+- Use the opportunity window to improve the API
+  - [Issue summarizing proposed API changes (#242)](https://github.com/cert-manager/trust-manager/issues/242)
+  - [Draft PR for proposed API changes done when migrating (#486)](https://github.com/cert-manager/trust-manager/pull/486)
+
+
+### Non-Goals
+
+- Behavior changes in `Bundle`/`ClusterBundle` API while migrating (only simpler API improvements)
+- Introduce a new API version
+- Introduce a namespace-scoped bundle API (at least not now)
+
+## Proposal
+
+We propose to do this change over multiple releases:
+
+1. Version X: Only `Bundle` API is served and has the functional controller. (present state)
+2. Version X+1: We introduce `ClusterBundle` with API otherwise identical to `Bundle`:
+   - `Bundle` is marked as deprecated.
+   - `ClusterBundle` introduced with agreed API improvements.
+   - Functional controller is migrated to `ClusterBundle`
+   - Validating webhook for both `Bundle` and `ClusterBundle` with same validations
+   - New technical controller added for `Bundle`, converting bundles to cluster bundles
+3. Version Y: Remove `Bundle` and only serve `ClusterBundle` (desired state)
+
+Some of these actions/states require extra attention, which are further described below.
+
+### Functional Controller Migration
+
+This should be rather simple. Since the APIs are identical, we just have to perform a "smart"
+search and replace in the controller code. When the controller initially boots up on the X+1 version,
+there will be no `ClusterBundle` resources in the cluster, as these will be created/updated by the technical
+controller.
+
+The `ClusterBundle` reconciler should do the same as for `Bundle` in version X, but we should test/verify that
+the controller is able to change owner references on target resources from `Bundle` to `ClusterBundle` without
+facing issues.
+
+We must also ensure that the `ClusterBundle` reconciler isn't deleting target resources it thinks are "orphaned".
+We believe the controller will only delete target resources if/when the controller resource exists, and
+otherwise rely on the Kubernetes garbage collector to delete orphans. But this should be tested/verified.
+
+### Temporary Technical Controller
+
+When the functional controller is migrated to target `ClusterBundle`, we want to introduce a temporary technical
+controller targeting the now deprecated `Bundle` resource.
+
+This controller should just create or update the `ClusterBundle` with the same name as the reconciled `Bundle`.
+Since the APIs should be quite similar, this should be a simple controller.
+
+- `Bundle` `.spec` should be copied (and eventually slightly converted) into `ClusterBundle` `.spec`
+- `Bundle` `.status` should be updated from `ClusterBundle` `.status`
```

Comment:
I think we should also add a clear message in the `Bundle` `.status` stating that using the `Bundle` is deprecated.
Also, the validation webhook should throw a warning explaining that the `Bundle` resources are deprecated.

Reply:
CRDs supports deprecating APIs/versions, and we can tailor the warning message. Isn't this enough?